src/protocol/requests/deleteRecords/v0/response.js

```diff
@@ -0,0 +1,51 @@
+const Decoder = require('../../../decoder')
+const { failure, createErrorFromCode } = require('../../../error')
+
+/**
+ * DeleteRecords Response (Version: 0) => throttle_time_ms [topics]
+ *  throttle_time_ms => INT32
+ *  topics => name [partitions]
+ *    name => STRING
+ *    partitions => partition low_watermark error_code
+ *      partition => INT32
+ *      low_watermark => INT64
+ *      error_code => INT16
+ */
+
+const topicNameComparator = (a, b) => a.topic.localeCompare(b.topic)
+
+const decode = async rawData => {
+  const decoder = new Decoder(rawData)
+  return {
+    throttleTime: decoder.readInt32(),
+    topics: decoder
+      .readArray(decoder => ({
+        topic: decoder.readString(),
+        partitions: decoder.readArray(decoder => ({
+          partition: decoder.readInt32(),
+          lowWatermark: decoder.readInt64(),
+          errorCode: decoder.readInt16(),
+        })),
+      }))
+      .sort(topicNameComparator),
+  }
+}
+
+const parse = async data => {
+  const topicsWithErrors = data.topics
+    .map(({ partitions }) => ({
+      partitionsWithErrors: partitions.filter(({ errorCode }) => failure(errorCode)),
+    }))
+    .filter(({ partitionsWithErrors }) => partitionsWithErrors.length)
+
+  if (topicsWithErrors.length > 0) {
+    throw createErrorFromCode(topicsWithErrors[0].partitionsWithErrors[0].errorCode)
```

Comment:
Aha, I just realised.. the "mapping-errors-to-partitions" issue is not a kafka problem, it's just a result of how this response is being parsed. I had used other parts of the code as an example, but forgot about it until now.. it's calling `createErrorFromCode()` which picks only the first available error, but I could add a similar helper that works on multiple errors and includes some metadata when there's more than 1 partition?

Means another tweak, but I think it might be worthwhile as it would help clarify which partition is wrong? Also I guess the way it's written now there's a small chance that it could drop an error if there are multiple ones, especially if they're not the same error (though I guess this is a very unlikely/pessimistic scenario 😂)